### ISSUE TITLE
0.6.7 breaks resumables with multiple fors over same variable

### ISSUE THREAD

**Issue by mweerden** [JuliaDynamics/ResumableFunctions.jl#86]
From version 0.6.7, the code below fails with `ERROR: LoadError: MethodError: Cannot `convert` an object of type String to an object of type Nothing`. It works fine with 0.6.6.

This happens because the same variable is used for multiple loops, and the type of that variable/slot in the generated iterator is determined by the last loop and only through code analysis, not the signature of `_empty`. This means that `s` is types as `Nothing`, causing the mentioned error in the first loop on actual assignment to `s`.

```
using ResumableFunctions

@resumable function singleton()::String
    @yield "anything"
end

# 1. must be defined before combined!
@resumable function _empty(x)::String
end

@resumable function combined()::String
    for s in singleton()
        @yield s # fails "here" because s is determined to be of type Nothing because of the second loop
    end
    # 2. must reuse of variable
    for s in _empty(1) # dummy argument needed to illustrate point 1.
        @yield s
    end
end

collect(combined())
```

The same happens with the (undocumented) `@yieldfrom`:
```
using ResumableFunctions

@resumable function singleton()::String
    @yield "anything"
end

@resumable function _empty()::String
end

@resumable function combined()::String
    # @yieldfrom also uses the same variable names for each generated loop
    @yieldfrom singleton()
    @yieldfrom _empty()
end

collect(combined())
```

**Comment by Krastanov** [comment]
Thank you for the report and apologize for the frustration this probably caused. 0.6.6 had some black magic related to inference that caused some bugs in edge cases so we had to remove it (for now). If you have the bandwidth, could you please consider reporting whether your code works on the following:

- ResumableFunctions 0.6.5, 0.6.6, and 0.6.7 (because it is important to see whether both 0.6.5 and 0.6.7 fail)
- each on Julia 1.9 and 1.10 (1.10 changed how some inference works in some of the versions)

Also, if you have a public library that depends on ResumableFunctions, let us know, we will add it to our CI to avoid unintentionally breaking it in the future.

**Comment by mweerden** [comment]
Yeah, took a bit of effort and head scratching to figure out. ;)

Quick testing of the above examples shows that it indeed only works with 0.6.6 (with both 1.9 and 1.10). I can try with the original code (unfortunately not public) next week, but don't expect different results.

**Comment by Krastanov** [comment]
In 0.6.6 we had a very talented contributor submit a much better inference/compiling pass for the resumable function macro. However, it had some minor regressions in performance that were not discovered until after julia 1.10 became public. We had to undo his contribution because of these regressions and he has not had a chance to work on it again since then (early January). It seems the functionality you needed was only available in that improvement from 0.6.6. We would like to get that improvement back in, but we are a bit starved for time right now, so probably it would not happen at least for a few more weeks.

**Comment by mweerden** [comment]
Sure, there is no rush. I've pinned it to 0.6.6 for now.

B.t.w., via #81 I saw that QuantumSavory/QuantumSavory.jl#89 seems to describe the same issue. So perhaps that package is a candidate for inclusion in your CI.

And here is a simpler example of the issue that can perhaps be included in your tests.
```
using ResumableFunctions

@resumable function f()
    for s in [""]
        @yield s
    end
    for s in [1]
        @yield s
    end
end

collect(f())
```

As far as I can tell, the issue is that `code_typed` can return multiple slots with the same name and in the current code it means only the type of the last will be used. It can be resolved by rereverting:
https://github.com/JuliaDynamics/ResumableFunctions.jl/blob/fdfbc515411219cc37dc36acfb8359cdfcc2b0f0/src/utils.jl#L61
and
https://github.com/JuliaDynamics/ResumableFunctions.jl/blob/fdfbc515411219cc37dc36acfb8359cdfcc2b0f0/src/utils.jl#L84

This works for me and does not reintroduce the other issue.

Do note that this still leaves the issue that `code_typed` infers that resumables always can yield `Nothing` or `Bool` (depending on Julia version), which practically is not the case. It adds to the confusion in the error.

Related to that, the documentation is a bit confusing on this topic as it kinda suggests that the return value of the function is also yielded. Might be a good idea the clarify that calling the "iterator" directly gives both the yielded and return values, while using it as a proper iterator, it only returns the yielded values.
